File: app/src/main/java/com/jgm90/cloudmusic/utils/Lyrics.java
package com.jgm90.cloudmusic.utils;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Lyrics {

    private static final String LRC_LINE_REGEXP = "\\[([0-9\\.:]+)](.*)";
    private static final String LRC_TIME_LONG_REGEXP = "^([0-9]{2}):([0-9]{2})\\.([0-9]{2})$";
    private static final String LRC_TIME_SHORT_REGEXP = "^([0-9]{2}):([0-9]{2})$";
    private static List<LyricLine> lyrics;
    private static double offset;

    public static List<LyricLine> parse(String lrc) {
        try {
            //separate lines
            lrc = lrc.replace("\n", "").replace("\r", "");
            String[] split = lrc.replace("[", "\n[").split("\n");
            split = Arrays.copyOfRange(split, 1, split.length - 1);


            //get offset tag if present
            Pattern offsetPattern = Pattern.compile("\\[offset: ([-0-9]+)].*");

            //get lines with valid tag
            lyrics = new ArrayList<LyricLine>();

            Pattern linePattern = Pattern.compile(LRC_LINE_REGEXP);
            for (String line : split) {
                Matcher lineMatcher = linePattern.matcher(line);

                if (lineMatcher.matches()) {
                    lyrics.add(new LyricLine(getSecondsFromTag(lineMatcher.group(1)), lineMatcher.group(2)));
                }

                Matcher offsetMatcher = offsetPattern.matcher(line);
                if (offsetMatcher.matches()) {
                    //get offset (ms)
                    offset = Double.parseDouble(offsetMatcher.group(1)) / 1000d;
                }
            }
            Collections.sort(lyrics, LyricLine.COMPARATOR);
            return lyrics;
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }

    private static double getSecondsFromTag(String tag) {
        double seconds = 0;
        if (tag.length() == 9) {
            //00:00.000
            seconds += Integer.parseInt(tag.substring(0, 2)) * 60;
            seconds += Integer.parseInt(tag.substring(3, 5));
            seconds += Integer.parseInt(tag.substring(6, 8)) / 100d;
            return seconds;
        } else if (tag.length() == 8) {
            //00:00.00
            seconds += Integer.parseInt(tag.substring(0, 2)) * 60;
            seconds += Integer.parseInt(tag.substring(3, 5));
            seconds += Integer.parseInt(tag.substring(6, 8)) / 100d;
            return seconds;
        } else if (tag.length() == 5) {
            //00:00
            seconds += Integer.parseInt(tag.substring(0, 2)) * 60;
            seconds += Integer.parseInt(tag.substring(3, 5));
            return seconds;
        } else throw new IllegalArgumentException("Not a valid time tag.");

    }

    public List<LyricLine> getAllLyrics() {
        return lyrics;
    }
}

File: app/src/main/java/com/jgm90/cloudmusic/utils/Lyrics.kt
package com.jgm90.cloudmusic.utils

import java.util.Arrays
import java.util.Collections
import java.util.regex.Pattern

class Lyrics {
    companion object {
        private const val LRC_LINE_REGEXP = "\\[([0-9\\.:]+)](.*)"
        private const val LRC_TIME_LONG_REGEXP = "^([0-9]{2}):([0-9]{2})\\.([0-9]{2})$"
        private const val LRC_TIME_SHORT_REGEXP = "^([0-9]{2}):([0-9]{2})$"
        private var lyrics: MutableList<LyricLine> = emptyList<LyricLine>().toMutableList()
        private var offset = 0.0

        @JvmStatic
        fun parse(lrc: String): List<LyricLine>? {
            var lrc = lrc
            try {
                //separate lines
                lrc = lrc.replace("\n", "").replace("\r", "")
                var split =
                    lrc.replace("[", "\n[").split("\n".toRegex()).dropLastWhile { it.isEmpty() }
                        .toTypedArray()
                split = Arrays.copyOfRange(split, 1, split.size - 1)


                //get offset tag if present
                val offsetPattern = Pattern.compile("\\[offset: ([-0-9]+)].*")

                //get lines with valid tag
                lyrics = ArrayList()

                val linePattern = Pattern.compile(LRC_LINE_REGEXP)
                for (line in split) {
                    val lineMatcher = linePattern.matcher(line)

                    if (lineMatcher.matches()) {
                        lyrics.add(
                            LyricLine(
                                getSecondsFromTag(lineMatcher.group(1)),
                                lineMatcher.group(2)
                            )
                        )
                    }

                    val offsetMatcher = offsetPattern.matcher(line)
                    if (offsetMatcher.matches()) {
                        //get offset (ms)
                        offset = offsetMatcher.group(1).toDouble() / 1000.0
                    }
                }
                Collections.sort(lyrics, LyricLine.COMPARATOR)
                return lyrics
            } catch (e: Exception) {
                e.printStackTrace()
                return null
            }
        }

        private fun getSecondsFromTag(tag: String): Double {
            var seconds = 0.0
            if (tag.length == 9) {
                //00:00.000
                seconds += (tag.substring(0, 2).toInt() * 60).toDouble()
                seconds += tag.substring(3, 5).toInt().toDouble()
                seconds += tag.substring(6, 8).toInt() / 100.0
                return seconds
            } else if (tag.length == 8) {
                //00:00.00
                seconds += (tag.substring(0, 2).toInt() * 60).toDouble()
                seconds += tag.substring(3, 5).toInt().toDouble()
                seconds += tag.substring(6, 8).toInt() / 100.0
                return seconds
            } else if (tag.length == 5) {
                //00:00
                seconds += (tag.substring(0, 2).toInt() * 60).toDouble()
                seconds += tag.substring(3, 5).toInt().toDouble()
                return seconds
            } else throw IllegalArgumentException("Not a valid time tag.")
        }
    }
}